docs/proposals/01-extensibility.md

```diff
@@ -0,0 +1,844 @@
+```
+
+##### Cloud config (user-data) for bootstrapping machines
+
+Gardener keeps keeping knowledge about the content of the cloud config scripts, but it will hand over it to the respective OS-specific controller which will generate the specific valid representation.
+Gardener creates two `MachineCloudConfig` CRDs, one for the cloud-config-downloader (which will later flow into the `WorkerPool` CRD) and one for the real cloud-config (which will be stored as a `Secret` in the Shoot's `kube-system` namespace, and downloaded and executed from the cloud-config-downloader on the machines).
+
+```yaml
+---
+apiVersion: extensions.gardener.cloud/v1alpha1
+kind: MachineCloudConfig
+metadata:
+  name: pool-01-downloader
+  namespace: shoot--core--aws-01
+spec:
+  type: CoreOS
+  units:
+  - name: cloud-config-downloader.service
+    command: start
+    enable: true
+    content: |
+      [Unit]
+      Description=Downloads the original cloud-config from Shoot API Server and execute it
+      After=docker.service docker.socket
+      Wants=docker.socket
+      [Service]
+      Restart=always
+      RestartSec=30
+      EnvironmentFile=/etc/environment
+      ExecStart=/bin/sh /var/lib/cloud-config-downloader/download-cloud-config.sh
+  files:
+  - path: /var/lib/cloud-config-downloader/kubeconfig
+    permissions: 0644
+    encoding: b64
+    content: LS0tCmFwaVZlcnNpb2...
+  - path: /var/lib/cloud-config-downloader/download-cloud-config.sh
+    permissions: 0644
+    encoding: b64
+    content: IyEvYmluL2Jhc2ggL...
+status:
+  observedGeneration: ...
+  state: ...
+  lastError: ..
+  lastOperation: ...
+  cloudConfig: | # base64-encoded
+    #cloud-config
+
+    coreos:
+      update:
+        reboot-strategy: off
+      units:
+      - name: cloud-config-downloader.service
+        command: start
+        enable: true
+        content: |
+          [Unit]
+          Description=Downloads the original cloud-config from Shoot API Server and execute it
+          After=docker.service docker.socket
+          Wants=docker.socket
+          [Service]
+          Restart=always
+          RestartSec=30
+          ...
+```
+
+:information: The cloud-config-downloader script does not only download the cloud-config initially but at regular intervals, e.g., every `30s`.
+If it sees an updated cloud-config then it applies it again by reloading and restarting all systemd units in order to reflect the changes.
+The way how this reloading of the cloud-config happens is OS-specific as well and not known to Gardener anymore, however, it must be part of the script already.
+On CoreOS, you have to execute `/usr/bin/coreos-cloudinit --from-file=<path>` whereas on SLES you execute `cloud-init --file <path> single -n write_files --frequency=once`.
+As Gardener doesn't know these commands it will write a placeholder expression instead (e.g., `{RELOAD-CLOUD-CONFIG-WITH-PATH:<path>}`) and the OS-specific controller is asked to replace it with the proper expression.
```

Comment:
Can we follow a similar approach for bootstrap tokens which can be replaced by MCM?

Reply:
Yes, we could to, but that requires interaction with MCM. See my reply to one of @mvladev's comments above: We still want to change the current bootstrap token logic, but let's do it separately and not within this proposal.